core/src/main/java/hudson/model/RunMap.java

```diff
@@ -256,6 +258,27 @@ protected boolean allowLoad(int buildNumber) {
         return true;
     }
 
+    @Restricted(NoExternalUse.class)
+    @Override
+    protected IntPredicate createLoadAllower() {
+        if (job == null) {
+            LOGGER.fine(() -> "deprecated constructor without Job used on " + dir);
+            return buildNumber -> true;
+        }
+        @SuppressWarnings("unchecked")
+        var allowers = RunListener.all().stream().map(l -> l.createLoadAllower(job)).toList();
```

Comment:
Could be optimized to exclude `RunListener`s which do not override the new method, though it is not clear the added reflection would actually be any faster. Or I suppose the default implementation could return null.